File: web/progress.mjs
const hints = {
  uploading: '歌曲只匯入這台電腦。',
  preparing: '正在讀取或轉換音訊。',
  recognizing: '依已處理的音訊回報，每段完成後更新。',
  cached: '重用這首歌先前的辨識結果。',
  aligning: '正在比對歌詞文字與聲音時間。',
  retrying: '這是目前補查片段的進度。',
  exporting: '正在儲存 SRT 與檢查表。',
};

export function formatProgress(progress = {}) {
  const { stage, percent } = progress;
  const value = Number.isFinite(percent) ? Math.max(0, Math.min(100, Math.round(percent))) : null;
  return {
    value,
    label: value === null ? '處理中' : `${value}%`,
    hint: hints[stage] || '完成目前階段後會繼續更新。',
  };
}

File: lyricflow/progress.py
"""The progress event contract shared by the CLI and background worker."""

import json

PROGRESS_PREFIX = "LYRIC_FLOW_PROGRESS "
STAGE_MESSAGES = {
    "preparing": "正在準備音訊…",
    "recognizing": "正在辨識歌曲…",
    "cached": "找到歌曲快取，已完成辨識。",
    "aligning": "正在對齊原始歌詞…",
    "retrying": "正在補查未對齊的句子…",
    "exporting": "正在產生字幕檔案…",
}


def report_progress(stage, percent=None, **details):
    print(PROGRESS_PREFIX + json.dumps({"stage": stage, "percent": percent, **details}), flush=True)


def progress_message(event):
    if event["stage"] == "retrying":
        if "end_line" in event:
            return (
                f"正在補辨識第 {event['current']}/{event['total']} 個片段"
                f"（歌詞第 {event['line']}–{event['end_line']} 行）…"
            )
        return f"正在補查第 {event['current']}/{event['total']} 句（歌詞第 {event['line']} 行）…"
    return STAGE_MESSAGES[event["stage"]]
